File: app/src/test/java/com/photoviewer/domain/usecases/SearchByTitleTest.java
package com.photoviewer.domain.usecases;


import com.photoviewer.data.entity.PhotoEntity;
import com.photoviewer.data.repository.PhotoEntityDataSource;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.Scheduler;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.verifyZeroInteractions;

public class SearchByTitleTest {
    private static final String FAKE_PHOTO_TITLE = "Fake Title";

    private SearchByTitle mSearchByTitle;

    @Mock
    private PhotoEntityDataSource mMockPhotoEntityDataSource;
    @Mock
    private Scheduler mMockScheduler;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        mSearchByTitle = new SearchByTitle(mMockScheduler,
                mMockScheduler, mMockPhotoEntityDataSource);
    }

    @Test
    public void testSearchByTitle() {
        List<PhotoEntity> photos = new ArrayList<PhotoEntity>() {{
            add(new PhotoEntity());
        }};
        given(mMockPhotoEntityDataSource.searchPhotosByTitle(FAKE_PHOTO_TITLE)).willReturn(Observable.just(photos));

        mSearchByTitle.setSearchedTitle(FAKE_PHOTO_TITLE);
        mSearchByTitle.buildObservable();

        verify(mMockPhotoEntityDataSource).searchPhotosByTitle(FAKE_PHOTO_TITLE);
        verifyNoMoreInteractions(mMockPhotoEntityDataSource);
        verifyZeroInteractions(mMockScheduler);
    }
}

File: app/src/test/java/com/photoviewer/domain/usecases/SearchByTitleTest.kt
@file:Suppress("IllegalIdentifier")
package com.photoviewer.domain.usecases

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import com.photoviewer.data.entity.PhotoEntity
import com.photoviewer.data.repository.PhotoEntityDataSource
import com.photoviewer.domain.Photo
import org.junit.Before
import org.junit.Test
import rx.Observable.just
import rx.Scheduler
import rx.observers.TestSubscriber
import java.util.*
import kotlin.test.assertEquals

class SearchByTitleTest {

    private val mMockPhotoEntityDataSource: PhotoEntityDataSource = mock()
    private val mMockScheduler: Scheduler = mock()

    private lateinit var mSearchByTitle: SearchByTitle
    private lateinit var mTestSubscriber: TestSubscriber<List<Photo>>

    companion object {
        private val FAKE_PHOTO_TITLE = "Fake Title"
    }

    @Before fun setUp() {
        mTestSubscriber = TestSubscriber.create()
        mSearchByTitle = SearchByTitle(mMockScheduler, mMockScheduler, mMockPhotoEntityDataSource)
    }

    @Test fun `should search by title`() {
        val photos = createPhotosList()

        assumeDataSourceHasSearchedContent(photos)

        mSearchByTitle.setSearchedTitle(FAKE_PHOTO_TITLE)
        mSearchByTitle.call().subscribe(mTestSubscriber)

        assertEquals(photos.size, mTestSubscriber.onNextEvents[0].size)
    }

    private fun assumeDataSourceHasSearchedContent(photos: ArrayList<PhotoEntity>) {
        whenever(mMockPhotoEntityDataSource.searchPhotosByTitle(FAKE_PHOTO_TITLE)).thenReturn(
                just<List<PhotoEntity>>(photos))
    }

    private fun createPhotosList() = ArrayList<PhotoEntity>().apply {
        add(PhotoEntity().apply { title = FAKE_PHOTO_TITLE })
    }
}